cli/program_state.rs

```diff
@@ -115,89 +119,69 @@ impl ProgramState {
   /// and traspilation.
   pub async fn prepare_module_load(
     self: &Arc<Self>,
-    module_specifier: ModuleSpecifier,
-    maybe_referrer: Option<ModuleSpecifier>,
+    specifier: ModuleSpecifier,
     target_lib: TargetLib,
     permissions: Permissions,
-    is_dyn_import: bool,
+    is_dynamic: bool,
     maybe_import_map: Option<ImportMap>,
   ) -> Result<(), AnyError> {
-    let module_specifier = module_specifier.clone();
+    let specifier = specifier.clone();
+    let handler = Rc::new(RefCell::new(FetchHandler::new(
+      self,
+      permissions.clone(),
+      self.permissions.clone(),
```

Comment:
This bit looks wrong to me - main permissions will be forwarded from the runtime while dynamic imports will use  initial permissions (which might have been partially revoked). And again, why do we need to different permissions passed to `FetchHandler`?

I've got this problem in the back of my head to remove `ProgramState::permissions` but waiting on these compiler refactors to be finished.

Reply:
Ok, I have looked at this... There is a lot of indirection in there that we should consider refactoring out.

- deno_core calls `prepare_load()` on the `ModuleLoader`
- `CliModuleLoader` implements `ModuleLoader`, and when `prepare_load()` is called it determines if the "main" module is being called, or if it is a dynamic import.
  - If it is the main module is passes `Permissions::all()` to `program_state.prepare_module_load()`,
  - Otherwise it clones a version of the permissions from `op_state`.
- The `program_state.permissions()` aren't used at all in figuring out dependencies as far as I can tell.

As stated above, the new module graph tries to make dynamic imports part of the graph, which gives a more complete type check when type checking modules, as well as an efficiency benefit of pre-fetching as many of the dynamic imports as possible.  Because of these seperate permissions being used in the new module graph, all the permission tests are passing.

I think I understand enough now to clean it up a bit better, and will take a look at the whole thing tomorrow and try to get it a bit more clean, since really it should be the module graph, being the centre of the universe, that mediates if it is using the "all" permissions, or the current runtime state permissions, and ideally not a concern of the module_loader or program_state.

Reply:
@kitsonk I agree that `ProgramState::permissions` should be removed. My only problem now is that `self.permissions.clone()` is the wrong one to pass to module graph. We can clean this up in a follow up PR, but in this case permissions from `OpState` should be passed. After that I can approve the PR

Reply:
Ok, to avoid any confusion, I renamed what gets passed in from the loader as `dynamic_permissions` and put several comments in the code.  Feels a lot clearer now.

Reply:
Thank you